From 9da3e9516e1ee36e8b6dd6a92946fa301daa9a91 Mon Sep 17 00:00:00 2001
From: "Subhransu S. Prusty" <subhransu.s.prusty@intel.com>
Date: Mon, 25 Aug 2014 14:11:55 +0530
Subject: [PATCH 05/12] Audio: New platform callback to disable and re-enable
 DAPM widgets
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

All the DAPM widgets are disabled and enabled for
stream recovery.The SST driver does this through a
callback registered with the platform.

Change-Id: I7f51ad02511913fa995b9eb0b6e07cbca6893bd8
Signed-off-by: Subhransu S. Prusty <subhransu.s.prusty@intel.com>
---
 sound/soc/intel/pcm.c          | 21 ++++++++++++++++-----
 sound/soc/intel/sst_platform.h |  1 +
 2 files changed, 17 insertions(+), 5 deletions(-)

diff --git a/sound/soc/intel/pcm.c b/sound/soc/intel/pcm.c
index b17db18..6fb1a4f 100644
--- a/sound/soc/intel/pcm.c
+++ b/sound/soc/intel/pcm.c
@@ -1021,14 +1021,16 @@ static int sst_platform_async_cb(struct sst_platform_cb_params *params)
 	struct snd_kcontrol *kcontrol;
 	struct sst_data *sst;
 
+	soc_platform = snd_soc_lookup_platform(sst_pdev);
+	if (!soc_platform) {
+		pr_err("Platform not found\n");
+		return -EINVAL;
+	}
+
 	switch (params->event) {
 	case SST_PLATFORM_VTSV_READ_EVENT: {
 		u8 *vtsv_result = params->params;
-		soc_platform = snd_soc_lookup_platform(sst_pdev);
-		if (!soc_platform) {
-			pr_err("Platform not found\n");
-			return -EINVAL;
-		}
+
 		sst = snd_soc_platform_get_drvdata(soc_platform);
 		card = soc_platform->card;
 		kcontrol = snd_soc_card_get_kcontrol(card, "vtsv event");
@@ -1042,6 +1044,15 @@ static int sst_platform_async_cb(struct sst_platform_cb_params *params)
 					&kcontrol->id);
 		break;
 	}
+
+	case SST_PLATFORM_TRIGGER_RECOVERY: {
+		bool *dapm_param = params->params;
+
+		card = soc_platform->card;
+		snd_soc_dapm_state_set(card, *dapm_param);
+		break;
+	}
+
 	default:
 		pr_info("No event handler for event Id %d\n", params->event);
 	}
diff --git a/sound/soc/intel/sst_platform.h b/sound/soc/intel/sst_platform.h
index c7b9447..6a218b0 100644
--- a/sound/soc/intel/sst_platform.h
+++ b/sound/soc/intel/sst_platform.h
@@ -150,6 +150,7 @@ struct sst_runtime_stream {
 };
 
 #define SST_PLATFORM_VTSV_READ_EVENT	0x1
+#define SST_PLATFORM_TRIGGER_RECOVERY	0x2
 struct sst_platform_cb_params {
 	/* Async event from firmware like VTSV*/
 	unsigned int event;
-- 
1.8.3.2

